### PR TITLE
Set UA string for Openverse API requests from the frontend

### PR DESCRIPTION
**Pull request by zackkrida**
## Description

Forgot I wrote this earlier in June. Adds a UA string identifying the frontend to the API in all requests. Useful for monitoring API traffic from the frontend over time.

## Testing Instructions

Run locally, make a client side search and check for the user agent set in the request in devtools.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [x] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: frontend/src/data/api-service.ts [hunk 1]
```diff
@@ -90,10 +92,15 @@ export const createApiService = ({
   const axiosParams: AxiosRequestConfig = {
     baseURL: isVersioned ? `${baseUrl}v1/` : baseUrl,
     timeout: DEFAULT_REQUEST_TIMEOUT,
+    headers: { "User-Agent": userAgent },
   }
+
   if (accessToken) {
-    axiosParams.headers = { Authorization: `Bearer ${accessToken}` }
+    if (axiosParams.headers) {
```
Comment (dhruvkb) [inline frontend/src/data/api-service.ts:99]: Why the `if`? Isn't `headers` hardcoded to be present on L95?

File: frontend/src/utils/user-agent.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/**
+ * A user agent string by which Openverse.org identifies itself to services
+ */
+const userAgent =
+  "Openverse/0.1 (https://openverse.org; openverse@wordpress.org)"
```
Comment (dhruvkb) [inline frontend/src/utils/user-agent.js:5]: Nit, potentially non-relevant: this version is rather arbitrary.

**Review by dhruvkb — APPROVED** [review]
LGTM!

File: frontend/src/utils/user-agent.js (hunk 2 above)
Comment (zackkrida) [reply frontend/src/utils/user-agent.js:5]: Agreed, I copied this from our usage somewhere else, but now I don't remember.  I'll try to find it and reassess an alternative

Comment (obulat) [reply frontend/src/utils/user-agent.js:5]: Nit: should this file go to the `/constants` folder instead?

**Review by obulat — COMMENTED** [review]
LGTM!

File: frontend/src/data/api-service.ts (hunk 1 above)
Comment (sarayourfriend) [reply frontend/src/data/api-service.ts:99]: The if is pretty confusing. When I first read it, it made me think that somehow `headers` could be undefined, even though it's _always_ on the axios params above.

There are two ways of doing this. One is purely in the type space and the other would be a slight change in the code to remove the if:

1. Change `const axiosParams: AxiosRequestConfig` to `const axiosParams: Required<Pick<AxiosRequestConfig, "headers">> & AxiosRequestConfig`. That tells TypeScript to remove `undefined` from the `headers` property, removing the need to check the `if` below. It's a bit wordy, but removes the need for unnecessary code. It also doesn't introduce the risk of a cast because if `headers` is _not_ defined on the right-hand side of the assignment expression, the type check will fail because the object won't fulfil the requirements of the type (`headers` being a headers object).
2. Cast headers below in the assignment to remove undefined: ``(axiosParams.headers as Exclude<typeof axiosParams["headers"], undefined>)["Authorization"] = `Bearer ${accessToken}` ``. To me this is objectively worse because if `headers` was removed from the `axiosParams` assignment above, there would be no type error.

I have a strong preference for the first option. To reduce the wordiness of it, you could create a new type:

```ts
/**
 * Openverse Axios request config with adjusted types for our use-case.
 */
type OpenverseAxiosRequestConfig = Required<Pick<AxiosRequestConfig, "headers">> & AxiosRequestConfig
```

Then use `OpenverseAxiosRequestConfig` in place of `AxiosRequestConfig` in the `axiosParams` assignment above.

File: frontend/src/utils/user-agent.js (hunk 2 above)
Comment (AetherUnbound) [reply frontend/src/utils/user-agent.js]: We use `0.1` for the catalog as well it seems: 

https://github.com/WordPress/openverse/blob/e1f48221938b4cc9838a85578b47bb4f6bfa33f6/catalog/dags/common/loader/provider_details.py#L121

**Review by AetherUnbound — COMMENTED** [review]
LGTM! I'm not seeing the user agent changed in the request headers when making a search locally though 🤔 it's the `/v1/images?q=<term>` request that the headers should be set, right?

**Comment by sarayourfriend** [comment]
@AetherUnbound IIRC Chrome does not allow user agent to be changed. Firefox also might not allow it. In that case, this change would only be evidenced for server side requests, not browser based ones. For Openverse.org requests from the frontend, the `referrer` header is a reliable identifier of source rather than the UA string.

**Comment by AetherUnbound** [comment]
Oh I see! Does that mean that `Referrer` here should be the value that was changed? Or is it as-expected?
image

**Comment by zackkrida** [comment]
@AetherUnbound Referrer is set by the browser for client-side requests, it's the url of whatever the site is, in this case it'll be https://openverse.org in production.

**Review by AetherUnbound — APPROVED** [review]
Thanks for the clarification!

**Comment by obulat** [comment]
> @AetherUnbound IIRC Chrome does not allow user agent to be changed. Firefox also might not allow it. In that case, this change would only be evidenced for server side requests, not browser based ones. For Openverse.org requests from the frontend, the `referrer` header is a reliable identifier of source rather than the UA string.

Is there a way to suppress the warnings? On staging, we get `instrument.js:289 Refused to set unsafe header "User-Agent"` on every client request now.

**Comment by zackkrida** [comment]
Ah, we should probably only set the UA server-side then. I'll make an issue today.
